cache/cache.go

```diff
@@ -0,0 +1,414 @@
+/*
+Copyright 2024 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"sync"
+	"time"
+)
+
+// Cache[T] is a thread-safe in-memory key/object store.
+// It can be used to store objects with optional expiration.
+type Cache[T any] struct {
+	*cache[T]
+	// keyFunc is used to make the key for objects stored in and retrieved from items, and
+	// should be deterministic.
+	keyFunc KeyFunc[T]
+}
+
+// Item is an item stored in the cache.
+type Item[T any] struct {
+	// Object is the item's object.
+	Object T
+	// Expiration is the item's expiration time.
+	Expiration int64
+}
+
+type cache[T any] struct {
+	// Items holds the elements in the cache.
+	Items map[string]Item[T]
+	// MaxItems is the maximum number of items the cache can hold.
+	MaxItems   int
+	metrics    *cacheMetrics
+	labelsFunc GetLvsFunc[T]
+	janitor    *janitor[T]
+	closed     bool
+
+	mu sync.RWMutex
+}
+
+var _ Expirable[any] = &Cache[any]{}
+
+// New creates a new cache with the given configuration.
+func New[T any](maxItems int, keyFunc KeyFunc[T], opts ...Options[T]) (*Cache[T], error) {
```

Comment:
Could we add `Dump` & `Load` functions so that the cache would be serialised to JSON on disk when the controller receives sigterm, and loaded into memory at startup?

Reply:
do we have a use case for persisting to disk? For the ones listed, values are either already persisted or can be requested and cached again with an api call (those are cheap as we are talking about credentials).

Reply:
I was thinking about image scanning and tag caching. For this we are using badger in IRC but maybe a simpler solution would suffice.

Reply:
This could be a good idea. What can be done here is optionally persist the cache in a file. The file would be periodically rotated to take into account the cache changes. The difference with badger is that we would have no LSM Tree implementation but a file that only grows. I think for IRC this is fine because the data is small, but this will have to be benchmarked. So It will have to be done in another PR.

Reply:
> a file that only grows

I don't think this is the case, the dump will contain whatever is in-memory when at controller shutdown. If keys have expired, the janitor would've removed them.

Reply:
yes, with expiration and eviction. Those are optional, but you are right, we have control.

Reply:
Same for LRU, the file will never grow beyond the limit

Reply:
for LRU, the capacity is in number of items. It doesn't limit the underlying file size. For expiration, if we set it to 0, its unlimited. We just need to be aware of the behavior anyway.